Feature/co 687 get csv record

gmtcrary opened the pull request:
WIP PR FOR CO-687

This feature branch adds CSVRecord handling for getEvents. It was branched of develop commit 9cfd56f82278ec5a8e1cd247aab487c0f1b2adfe.

As of 62d402b the CSV record gets written to a events_gospel[TIMESTAMP].csv file locally. This will be changed before check in.

TODO: Handle instances, customfields, filters.

Commit:
Add opencsv dependancy to build.gradle

Commit:
Add hashbang to cloneContribFiles.sh

Commit:
Add initial csv record handling code to EventsResource getEvents method

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -67,13 +68,51 @@ class EventsResource extends Resource {
+                              "Ticket URL" , "Cost" , "Hashtag" , "Facebook URL" , "Group" ,
+                              "Department" , "Allow User Activity" , "Allow User Attendance" ,
+                              "Visibility" , "Featured Tabs" , "Sponsored" , "Venue Page Only" ,
+                              "Exclude From Trending" , "Event Types" , "Top level filter" ,
```
We don't want to hard code any filters or custom fields. These things will differ depending on what we have stored in our local cache. Localist expects each custom field and filter to be it's own column. You can use [this DAO method ](https://github.com/osu-mist/osu-events-api/blob/develop/src/main/groovy/edu/oregonstate/mist/osuevents/db/EventsDAO.groovy#L276-L278) to get all custom fields from the local cache and [this DAO method](https://github.com/osu-mist/osu-events-api/blob/develop/src/main/groovy/edu/oregonstate/mist/osuevents/db/EventsDAO.groovy#L340-L342) to get filters from the local cache.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -67,13 +68,51 @@ class EventsResource extends Resource {
+            CSVWriter writer = new CSVWriter(new FileWriter(csvfilename))
+
+            //CSV Layout Header
+            writer.writeNext(["Title" , "Description" , "Date From" , "Date To" , "Recurrence" ,
```
Don't forget about "External ID" column

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -67,13 +68,51 @@ class EventsResource extends Resource {
+                writer.writeNext(event.toCSVRecord())
+            }
+
+            writer.close()
```
I'm thinking we should move all CSV related stuff to a different class. Then the code for the Events Resource would be something like:
`if (format == "csv") {
    def eventsCSV = resultObjectToCSV(resultObject) }`

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -67,13 +68,51 @@ class EventsResource extends Resource {
  */
     @GET
     @Produces(MediaType.APPLICATION_JSON)
+    //TODO Add CSV MediaType to response object
```
@Produces({"text/csv"})

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -67,13 +68,51 @@ class EventsResource extends Resource {
+                              "Exclude From Trending" , "Event Types" , "Top level filter" ,
+                              "Custom Field"] as String[])
+
+            resultObject.data.each {
```
remember, each row in the CSV will be an instance, so this closure will be like
```
resultObject.data.each {
    it.attributes.instances.each {
        writeacsvline()
    }
}
```

Commit:
Merge branch 'develop' into feature/CO-687-GET-CSV-RECORD

Commit:
created helper function for formatting dates and times for csv

Commit:
consistent quotes

Commit:
timezone as a config value

the timezone for the csv endpoint (the only endpoint that won't return times in UTC) depends on the environment, thus set it in configuration.yaml

Commit:
make it DRY

Commit:
added unit test for csv time methods

Commit:
merge from develop, add default timezone

added default timezone as UTC for EventsResource

jaredkosanovic committed:
Merge pull request #5 from osu-mist/feature/CO-732-time-formatting-csv

Feature/co 732 time formatting csv

Commit:
Add WIP instance handling

Commit:
Add WIP Custom Fields handling

Commit:
Add csv MediaType to getEvents produces annotation

Commit:
Add filter handling to getEvents CSV response

Commit:
Remove unused and unfinished toCSVRecord method in Event class

Commit:
Add csv response entity to getEvents method and handle file cleanup

Commit:
Add comments to filters and custom fields mappings

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
     @GET
-    @Produces(MediaType.APPLICATION_JSON)
+    @Produces([MediaType.APPLICATION_JSON, "text/csv"])
+    //TODO Add CSV MediaType to response object
```
Old todo?

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+                    }
+                }
+
+                //Map over the custom field entries in the same order as they were added to the header
```
Codenarc thinks this line is too long.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+
+        if (!format) {
+            events.each {
+                resultObject.data += new ResourceObject(
```
resultObject wasn't initiated as a `new ResultObject` so when a I make a request for /events, it's throwing a null pointer exception error.

Also, the way you have this setup, it's not adding any instances to each resource object. Why did you add logic for this? Can we use the getResultObject method instead? Here's what I'm thinking:
```
def events = eventsDAO.getEvents()
ResultObject allEvents = getResultObject(events)
if (!format) {
    return Response.ok(allEvents).type(MediaType.APPLICATION_JSON).build()
} else if (format == "csv" {
    //the rest of the stuff

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+                        attributes: it
+                )
+            }
+        } else if (format == "csv") {
```
This is too much code for this one method. Could you move all CSV manipulation code to CSVHelperFunctions.groovy? It could be split up into several static methods (public or private as needed), then for this EventsResource, it could just call one method to get a CSV. All the csv method(s) should need is a ResultObject with all events and instances.

```
} else if (format == "csv") {
    def allEventsCsv = CSVHelperFunctions.getCSV(allEvents)
    return Response.ok(allEventsCsv, "text/csv").build()
}

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+                    baseEventRecord.add(entryVal)
+                }
+
+                List<Instance> eventInstances = eventsDAO.getInstances(eventid)
```
Since you used the getResultObject method, you shouldn't have to call the eventsDAO.getInstances method since each ResourceObject in the ResultObject already contains instances.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+                )
+            }
+        } else if (format == "csv") {
+            final ArrayList<String> BASECSVHEADER =
```
We aren't using "Allow User Attendance", "Allow User Activity", or "Featured Tabs", we should remove those from the csv header and remove the placeholder values being populated for each event.

jaredkosanovic replied:
We can add those fields later.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+            }
+        } else if (format == "csv") {
+            final ArrayList<String> BASECSVHEADER =
+                    ["EventID","Title" , "Description" , "Date From" , "Date To" ,
```
Can we change "EventID" to "External ID" ? Although Localist removed reference if this from their docs, this is what field their expecting for unique ID's.

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/osuevents/resources/EventsResource.groovy:
```diff
@@ -66,14 +74,116 @@ class EventsResource extends Resource {
+                    String[] instanceRecord = baseEventRecord.clone()
+
+                    //Replaces "DATE_FROM_PLACE_HOLDER"
+                    instanceRecord[3] = CSVHelperFunctions.getCSVDate(it.start,backendTimezone)
```
If I had to add one field to the csv, let's say in the second column, all of these instanceRecord indexes would have to change, and I would need to change baseEventRecord as well as BASECSVHEADER. I'm thinking we should maintain the headers in such a way that we can reference their positions. That way we wouldn't have to worry about keeping this in sync/order.

jaredkosanovic replied:
You could use something like indexOf() to get the numbered index of the field you want instead of hardcoding the index number.

jaredkosanovic reviewed (changes requested):
Left some comments.

Commit:
Remove old todo about CSV mediatype

Commit:
Remove redundant getResultObject work when handling json response

Commit:
Fix comment line length for codenarc

Commit:
Add WIP refactor of CSV creation to CSVHelperFunctions.groovy
